README.md
# in_app_review

[![pub package](https://img.shields.io/pub/v/in_app_review.svg)](https://pub.dartlang.org/packages/in_app_review) [![pub points](https://badges.bar/in_app_review/pub%20points)](https://pub.dev/packages/in_app_review/score) [![likes](https://badges.bar/in_app_review/likes)](https://pub.dev/packages/in_app_review/score) [![popularity](https://badges.bar/in_app_review/popularity)](https://pub.dev/packages/in_app_review/score)

![In-App Review Android Demo](https://github.com/britannio/in_app_review/blob/master/screenshots/android.jpg)
![In-App Review IOS Demo](https://github.com/britannio/in_app_review/blob/master/screenshots/ios.png)

# Description
Flutter plugin that lets you show a system review pop up where users can leave a review for your app without needing to leave it. Alternatively you can open your store listing via a deep link.

Uses the [In-App Review](https://developer.android.com/guide/playcore/in-app-review) API on Android and the [SKStoreReviewController](https://developer.apple.com/documentation/storekit/skstorereviewcontroller) on iOS/MacOS


# Usage
Use the following code to attempt to display the system review pop up. Beware that this does **not** guarantee that the pop up will be displayed as the quota may have been exceeded. On IOS & MacOS, this quota is three times in a 365 day period. On Android it is unclear: https://developer.android.com/guide/playcore/in-app-review#quotas
```dart
import 'package:in_app_review/in_app_review.dart';

final InAppReview inAppReview = InAppReview.instance;

if (await inAppReview.isAvailable()) {
    inAppReview.requestReview();
}
```

Use the following code to open the Google Play Store on Android or the App Store on IOS & MacOS. This is the recommended option if you want to permanently provide a button or other call-to-action to let users leave a review.
```dart
import 'package:in_app_review/in_app_review.dart';

final InAppReview inAppReview = InAppReview.instance;

inAppReview.openStoreListing(appStoreId: '<YOUR_APP_STORE_ID>');
```

`appStoreId` is only required on IOS and MacOS and can be found in App Store Connect under General > App Information > Apple ID

# Guidelines
https://developer.apple.com/design/human-interface-guidelines/ios/system-capabilities/ratings-and-reviews/

https://developer.android.com/guide/playcore/in-app-review#when-to-request
https://developer.android.com/guide/playcore/in-app-review#design-guidelines

Since there is a quota on how many times the pop up can be shown, you should **not** trigger `requestReview()` via a button or other *call-to-action* option. Instead, you can reliably redirect users to your store listing via `openStoreListing()`.


# Requirements
## Android
Requires Android 5 Lollipop(API 21) or higher and the Google Play Store must be installed.
## IOS
Requires IOS version 10.3
## MacOS
Requires MacOS version 10.14

# Testing
## Android
You must upload your app to the Play Store to test `requestReview()`. An easy way to do this is to build an apk/app bundle and upload it via [internal app sharing](https://play.google.com/apps/publish/internalappsharing/).

More details at https://developer.android.com/guide/playcore/in-app-review/test

## IOS
This plugin can be tested via the IOS simulator or on a physical device. 
Note that `requestReview()` has no effect when testing via TestFlight [as documented](https://developer.apple.com/documentation/storekit/skstorereviewcontroller/2851536-requestreview#discussion).

## MacOS
This plugin can be tested by running your MacOS application locally.


Issues & pull requests are more than welcome!

CHANGELOG.md
# [0.2.1+1]

- Improve IOS testing docs.

# [0.2.1]

- Update dependencies.
- Android Play Core Library V1.8.2 release notes:
    - Fixed UI flickering in the In-App Review API

# [0.2.0+4]

- Remove deprecated API warning.
- Update dependencies.

# [0.2.0+3]

- Instructions in the README have been improved along with the example.

# [0.2.0+2]

- Update changelog format

# [0.2.0+1]

- Update MacOS testing instructions

# [0.2.0] Breaking Change

- Add MacOS support
- Rename `openStoreListing(iOSAppStoreId: '')` to `openStoreListing(appStoreId: '')` 

# [0.1.0]

- Improve docs
- Set Android minSdkVersion to 16
- Refactor Android Plugin

# [0.0.1]

Initial release


in_app_review/lib/in_app_review.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/services.dart';
import 'package:package_info/package_info.dart';
import 'package:url_launcher/url_launcher.dart';

class InAppReview {
  InAppReview._();

  static final InAppReview instance = InAppReview._();

  static const MethodChannel _channel =
      const MethodChannel('dev.britannio.in_app_review');

  /// Check's if the device is able to show a review dialog.
  ///
  /// On Android the Google Play Store must be installed and the device must be
  /// running **Android 5 Lollipop(API 21)** or higher.
  ///
  /// IOS devices must be running **IOS version 10.3** or higher.
  ///
  /// MacOS devices must be running **MacOS version 10.14** or higher
  Future<bool> isAvailable() => _channel.invokeMethod('isAvailable');

  /// Attempts to show the review dialog. It's recommended to first check if
  /// this cannot be done via [isAvailable]. If it is not available then
  /// you can open the store listing via [openStoreListing].
  ///
  /// To improve the users experience, IOS and Android enforce limitations
  /// that might prevent this from working after a few tries. IOS & MacOS users
  /// can also disable this feature entirely in the App Store settings.
  ///
  /// More info and guidance:
  /// https://developer.android.com/guide/playcore/in-app-review#when-to-request
  /// https://developer.apple.com/design/human-interface-guidelines/ios/system-capabilities/ratings-and-reviews/
  /// https://developer.apple.com/design/human-interface-guidelines/macos/system-capabilities/ratings-and-reviews/
  Future<void> requestReview() => _channel.invokeMethod('requestReview');

  /// Opens the Play Store on Android and opens the App Store with a review
  /// screen on IOS & MacOS. [appStoreId] is required for IOS & MacOS.
  Future<void> openStoreListing({
    /// Required for IOS & MacOS
    String appStoreId,
  }) async {
    final bool isIOS = Platform.isIOS;
    final bool isMacOS = Platform.isMacOS;
    final bool isAndroid = Platform.isAndroid;

    if (isIOS || isMacOS) {
      assert(appStoreId != null);

      final Uri uri = Uri(
        scheme: isIOS ? 'https' : 'macappstore',
        host: 'apps.apple.com',
        path: 'app/id$appStoreId',
        queryParameters: {'action': 'write-review'},
      );

      await _launchUrl(uri.toString());
    } else if (isAndroid) {
      final PackageInfo packageInfo = await PackageInfo.fromPlatform();

      final String packageName = packageInfo.packageName;

      await _launchUrl(
        'https://play.google.com/store/apps/details?id=$packageName',
      );
    } else {
      throw UnsupportedError('Platform not supported');
    }
  }

  Future<void> _launchUrl(String url) async {
    if (await canLaunch(url)) {
      await launch(url);
    }
  }
}
